### PR TITLE
perf(cdk/overlay): do not run change detection if there are no `backdropClick` listeners

### PR DESCRIPTION
**Pull request by arturovt**

### CODE REVIEW

File: src/cdk/overlay/overlay.spec.ts [hunk 1]
```diff
@@ -1065,6 +1066,65 @@ describe('Overlay', () => {
   });
 });
 
+// Caretaker note: this is moved to a separate `describe` block because the first `describe` block
```
Comment (jelbourn) [inline src/cdk/overlay/overlay.spec.ts:1069]: Saying "caretaker note" here isn't necessary, this is just a regular code comment. We use "Caretaker note" in PR comments to mention something that might come up during the process of merging the PR and syncing it into Google's internal code repo.

File: src/cdk/overlay/overlay-ref.ts [hunk 2]
```diff
@@ -381,7 +385,9 @@ export class OverlayRef implements PortalOutlet, OverlayReference {
 
     // Forward backdrop clicks such that the consumer of the overlay can perform whatever
     // action desired when such a click occurs (usually closing the overlay).
-    this._backdropElement.addEventListener('click', this._backdropClickHandler);
+    this._ngZone.runOutsideAngular(() => {
```
Comment (jelbourn) [inline src/cdk/overlay/overlay-ref.ts:388]: Could this change be decoupled from the length check above? I suspect we'll run into a number of apps that depend on change detection from this event

Comment (arturovt) [reply src/cdk/overlay/overlay-ref.ts:388]: Hey, could you elaborate a little bit on that? E.g. some code example?

Comment (jelbourn) [reply src/cdk/overlay/overlay-ref.ts:388]: It's not something it's easy to create a code example for, unfortunately. We often run into a problem when landing changes like this into Google where some team's unit tests are implicitly depending on specific number and sequence of change detections in order for their assertions to pass. Removing change detection from something that used to have one almost always makes some number of tests fail.

**Pull request closed without merging**